source/stylist/fortran.py

```diff
@@ -392,3 +393,67 @@ def examine_fortran(self, subject: FortranSource) -> List[Issue]:
 
         issues.sort(key=lambda x: (x.filename, x.line, x.description))
         return issues
+
+
+class AutoCharArrayIntent(FortranRule):
+    """
+    Checks that all automatically assigned character arrays used as
+    subroutine or function arguments have intent(in) to avoid writing
+    outside the given array.
+    """
+    def __init__(self):
+        pass
+
+    def message(self, name, intent):
+        return (f"Arguments of type character(*) must have intent IN, but "
+                f"{name} has intent {intent}.")
+
+    def examine_fortran(self, subject: FortranSource) -> List[Issue]:
+        issues = []
+
+        # Collect all variable declarations
+        declarations: List[Fortran2003.Base] = list(
+            subject.find_all(Fortran2003.Type_Declaration_Stmt)
+        )
+
+        # keep only the ones in subroutines
+        declarations = [
+            declaration
+            for declaration
+            in declarations
+            if isinstance(
+                declaration.parent.parent,
+                Fortran2003.Subroutine_Subprogram
+            )
+        ]
+
+        for declaration in declarations:
+            type_spec = declaration.items[0]
+            # If not a character, no concern
+            if not type_spec.items[0] == "CHARACTER":
+                continue
+            param_value = type_spec.items[1]
+            # This might be a length selector, if so get the param value
+            if isinstance(param_value, Fortran2003.Length_Selector):
+                param_value = param_value.items[1]
```

Comment:
What happens when this test fails? Does `param_value` still contain something valid for the next test?

Reply:
A type spec like `character(*)` parses as `type_spec.param_value` but a spec like `character(len=*)` parses as `type_spec.length_selector.param_value`.
This if clause catches the latter case and skips the extra layer of indirection. After this point `param_value` definitely contains the array length.
The test case includes examples of both styles.